### PR TITLE
Improved discovery time [4661]

### PR DESCRIPTION
**Pull request by richiware**
Improved the time needed for discovering remote participants and theirs entities.

- Reconfigured the attributes of EDP endpoints.
- Using shrinked locators in several places not used yet.

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1222/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1192/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1205/)

File: include/fastrtps/rtps/writer/timedevent/NackResponseDelay.h [hunk 1]
```diff
@@ -35,26 +35,31 @@ class ReaderProxy;
...
+class NackResponseDelay:public TimedEvent
+{
+    public:
+
+        /*!
+         *
+         * @param[in] writer Writer which creates this event.
+         * @param[in] interval_millisec Interval of the event in milliseconds.
+         */
+        NackResponseDelay(StatefulWriter* writer, double interval_millisec);
```
Comment (MiguelCompany) [inline include/fastrtps/rtps/writer/timedevent/NackResponseDelay.h]: New styling requires parameters on different lines

File: include/fastrtps/rtps/writer/timedevent/NackResponseDelay.h [hunk 2]
```diff
@@ -35,26 +35,31 @@ class ReaderProxy;
...
+
+        virtual ~NackResponseDelay();
+
+        /*!
+         * Method invoked when the event occurs
+         *
+         * @param code Code representing the status of the event
+         * @param msg Message associated to the event
+         */
+        void event(EventCode code, const char* msg= nullptr);
```
Comment (MiguelCompany) [inline include/fastrtps/rtps/writer/timedevent/NackResponseDelay.h]: New styling requires parameters on different lines

File: include/fastrtps/rtps/writer/timedevent/NackResponseDelay.h [hunk 3]
```diff
@@ -35,26 +35,31 @@ class ReaderProxy;
...
+
+    private:
+
+        //!Associated reader proxy
```
Comment (MiguelCompany) [inline include/fastrtps/rtps/writer/timedevent/NackResponseDelay.h]:
```suggestion
        //!Associated writer
```

File: src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp [hunk 4]
```diff
@@ -153,10 +153,8 @@ bool EDPSimple::createSEDPEndpoints()
         watt.endpoint.multicastLocatorList = this->mp_PDP->getLocalParticipantProxyData()->m_metatrafficMulticastLocatorList;
         //watt.endpoint.remoteLocatorList = m_discovery.initialPeersList;
         watt.endpoint.durabilityKind = TRANSIENT_LOCAL;
-        watt.times.nackResponseDelay.seconds = 0;
-        watt.times.nackResponseDelay.fraction = 0;
-        watt.times.initialHeartbeatDelay.seconds = 0;
-        watt.times.initialHeartbeatDelay.fraction = 0;
+        watt.times.heartbeatPeriod.seconds = 1;
+        watt.times.heartbeatPeriod.fraction = 0;
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp]: Maybe use specific values for all the watt.times fields? Even better, have a const for it at the top of the file?

File: src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp [hunk 5]
```diff
@@ -184,10 +182,6 @@ bool EDPSimple::createSEDPEndpoints()
         ratt.endpoint.multicastLocatorList = this->mp_PDP->getLocalParticipantProxyData()->m_metatrafficMulticastLocatorList;
         //ratt.endpoint.remoteLocatorList = m_discovery.initialPeersList;
         ratt.endpoint.durabilityKind = TRANSIENT_LOCAL;
-        ratt.times.heartbeatResponseDelay.seconds = 0;
-        ratt.times.heartbeatResponseDelay.fraction = 0;
-        ratt.times.initialAcknackDelay.seconds = 0;
-        ratt.times.initialAcknackDelay.fraction = 0;
         this->mp_subListen = new EDPSimpleSUBListener(this);
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp]: Maybe use specific values for all the ratt.times fields? Even better, have a const for it at the top of the file?

File: src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp [hunk 6]
```diff
@@ -218,10 +212,6 @@ bool EDPSimple::createSEDPEndpoints()
         ratt.endpoint.multicastLocatorList = this->mp_PDP->getLocalParticipantProxyData()->m_metatrafficMulticastLocatorList;
         //ratt.endpoint.remoteLocatorList = m_discovery.initialPeersList;
         ratt.endpoint.durabilityKind = TRANSIENT_LOCAL;
-        ratt.times.heartbeatResponseDelay.seconds = 0;
-        ratt.times.heartbeatResponseDelay.fraction = 0;
-        ratt.times.initialAcknackDelay.seconds = 0;
-        ratt.times.initialAcknackDelay.fraction = 0;
         this->mp_pubListen = new EDPSimplePUBListener(this);
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp]: Maybe use specific values for all the ratt.times fields? Even better, have a const for it at the top of the file?

File: src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp [hunk 7]
```diff
@@ -249,10 +239,8 @@ bool EDPSimple::createSEDPEndpoints()
         watt.endpoint.multicastLocatorList = this->mp_PDP->getLocalParticipantProxyData()->m_metatrafficMulticastLocatorList;
         //watt.endpoint.remoteLocatorList = m_discovery.initialPeersList;
         watt.endpoint.durabilityKind = TRANSIENT_LOCAL;
-        watt.times.nackResponseDelay.seconds = 0;
-        watt.times.nackResponseDelay.fraction = 0;
-        watt.times.initialHeartbeatDelay.seconds = 0;
-        watt.times.initialHeartbeatDelay.fraction = 0;
+        watt.times.heartbeatPeriod.seconds = 1;
+        watt.times.heartbeatPeriod.fraction =0;
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp]: Maybe use specific values for all the watt.times fields? Even better, have a const for it at the top of the file?

File: src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp [hunk 8]
```diff
@@ -301,10 +289,8 @@ bool EDPSimple::create_sedp_secure_endpoints()
         watt.endpoint.multicastLocatorList = this->mp_PDP->getLocalParticipantProxyData()->m_metatrafficMulticastLocatorList;
         //watt.endpoint.remoteLocatorList = m_discovery.initialPeersList;
         watt.endpoint.durabilityKind = TRANSIENT_LOCAL;
-        watt.times.nackResponseDelay.seconds = 0;
-        watt.times.nackResponseDelay.fraction = 0;
-        watt.times.initialHeartbeatDelay.seconds = 0;
-        watt.times.initialHeartbeatDelay.fraction = 0;
+        watt.times.heartbeatPeriod.seconds = 1;
+        watt.times.heartbeatPeriod.fraction = 0;
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp]: Maybe use specific values for all the watt.times fields? Even better, have a const for it at the top of the file?

File: src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp [hunk 9]
```diff
@@ -343,10 +329,6 @@ bool EDPSimple::create_sedp_secure_endpoints()
         ratt.endpoint.multicastLocatorList = this->mp_PDP->getLocalParticipantProxyData()->m_metatrafficMulticastLocatorList;
         //ratt.endpoint.remoteLocatorList = m_discovery.initialPeersList;
         ratt.endpoint.durabilityKind = TRANSIENT_LOCAL;
-        ratt.times.heartbeatResponseDelay.seconds = 0;
-        ratt.times.heartbeatResponseDelay.fraction = 0;
-        ratt.times.initialAcknackDelay.seconds = 0;
-        ratt.times.initialAcknackDelay.fraction = 0;
         ratt.endpoint.security_attributes().is_submessage_protected = part_attr.is_discovery_protected;
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp]: Maybe use specific values for all the ratt.times fields? Even better, have a const for it at the top of the file?

File: src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp [hunk 10]
```diff
@@ -385,10 +367,6 @@ bool EDPSimple::create_sedp_secure_endpoints()
...
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp]: Maybe use specific values for all the ratt.times fields? Even better, have a const for it at the top of the file?

File: src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp [hunk 11]
```diff
@@ -423,10 +401,8 @@ bool EDPSimple::create_sedp_secure_endpoints()
         watt.endpoint.multicastLocatorList = this->mp_PDP->getLocalParticipantProxyData()->m_metatrafficMulticastLocatorList;
         //watt.endpoint.remoteLocatorList = m_discovery.initialPeersList;
         watt.endpoint.durabilityKind = TRANSIENT_LOCAL;
-        watt.times.nackResponseDelay.seconds = 0;
-        watt.times.nackResponseDelay.fraction = 0;
-        watt.times.initialHeartbeatDelay.seconds = 0;
-        watt.times.initialHeartbeatDelay.fraction = 0;
+        watt.times.heartbeatPeriod.seconds = 1;
+        watt.times.heartbeatPeriod.fraction = 0;
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp]: Maybe use specific values for all the watt.times fields? Even better, have a const for it at the top of the file?

File: src/cpp/rtps/reader/timedevent/InitialAckNack.cpp [hunk 12]
```diff
@@ -51,7 +51,10 @@ InitialAckNack::InitialAckNack(WriterProxy* wp, double interval):
     m_destination_locators(wp->m_att.endpoint.unicastLocatorList),
```
Comment (MiguelCompany) [inline src/cpp/rtps/reader/timedevent/InitialAckNack.cpp]: Shrink locators?

File: src/cpp/rtps/writer/ReaderProxy.cpp [hunk 13]
```diff
@@ -35,13 +34,11 @@ using namespace eprosima::fastrtps::rtps;
 
 
 ReaderProxy::ReaderProxy(const RemoteReaderAttributes& rdata,const WriterTimes& times,StatefulWriter* SW) :
-    m_att(rdata), mp_SFW(SW),
-    mp_nackResponse(nullptr), mp_nackSupression(nullptr), m_lastAcknackCount(0),
+    m_att(rdata), mp_SFW(SW), mp_nackSupression(nullptr), m_lastAcknackCount(0),
```
Comment (MiguelCompany) [inline src/cpp/rtps/writer/ReaderProxy.cpp]: Please follow new coding style rules

File: src/cpp/rtps/writer/StatefulWriter.cpp [hunk 14]
```diff
@@ -78,6 +80,12 @@ StatefulWriter::~StatefulWriter()
             it != matched_readers.end(); ++it)
         (*it)->destroy_timers();
```
Comment (MiguelCompany) [inline src/cpp/rtps/writer/StatefulWriter.cpp]: Style here (braces)

File: src/cpp/rtps/writer/StatefulWriter.cpp [hunk 15]
```diff
@@ -78,6 +80,12 @@ StatefulWriter::~StatefulWriter()
...
 
+    if(nack_response_event_ != nullptr)
+    {
+        delete(nack_response_event_);
+        nack_response_event_ = nullptr;
+    }
+
     if(mp_periodicHB !=nullptr)
         delete(mp_periodicHB);
```
Comment (MiguelCompany) [inline src/cpp/rtps/writer/StatefulWriter.cpp]: Style here (braces)

File: src/cpp/rtps/writer/StatefulWriter.cpp [hunk 16]
```diff
@@ -142,13 +150,16 @@ void StatefulWriter::unsent_change_added_to_history(CacheChange_t* change)
                 if (m_separateSendingEnabled)
                 {
                     guids.at(0) = (*it)->m_att.guid;
-                    RTPSMessageGroup group(mp_RTPSParticipant, this, RTPSMessageGroup::WRITER, m_cdrmessages, 
-                        (*it)->m_att.endpoint.remoteLocatorList, guids);
-                    if (!group.add_data(*change, guids, (*it)->m_att.endpoint.remoteLocatorList, (*it)->m_att.expectsInlineQos))
+                    LocatorList_t remote_locators_shrinked{mp_RTPSParticipant->network_factory().ShrinkLocatorLists(
+                            {(*it)->m_att.endpoint.remoteLocatorList})};
+
```
Comment (MiguelCompany) [inline src/cpp/rtps/writer/StatefulWriter.cpp]: If I remeber correctly, remoteLocatorList is shrinked inside `matched_reader_add`, so there is no need to shrink it here

File: src/cpp/rtps/writer/StatefulWriter.cpp [hunk 17]
```diff
@@ -242,8 +253,10 @@ void StatefulWriter::send_any_unsent_changes()
 
             // Specific destination message group
             guids.at(0) = remoteReader->m_att.guid;
-            const LocatorList_t& locators = remoteReader->m_att.endpoint.remoteLocatorList;
-            RTPSMessageGroup group(mp_RTPSParticipant, this, RTPSMessageGroup::WRITER, m_cdrmessages, locators, guids);
+            const LocatorList_t remote_locators_shrinked{mp_RTPSParticipant->network_factory().ShrinkLocatorLists(
+                    {remoteReader->m_att.endpoint.remoteLocatorList})};
+            RTPSMessageGroup group(mp_RTPSParticipant, this, RTPSMessageGroup::WRITER, m_cdrmessages,
```
Comment (MiguelCompany) [inline src/cpp/rtps/writer/StatefulWriter.cpp]: If I remeber correctly, remoteLocatorList is shrinked inside `matched_reader_add`, so there is no need to shrink it here

File: src/cpp/rtps/writer/StatefulWriter.cpp [hunk 18]
```diff
@@ -558,16 +572,18 @@ bool StatefulWriter::matched_reader_add(RemoteReaderAttributes& rdata)
         assert(last_seq + 1 == current_seq);
 
         std::vector<GUID_t> guids(1, rp->m_att.guid);
-        const LocatorList_t& locatorsList = rp->m_att.endpoint.remoteLocatorList;
-        RTPSMessageGroup group(mp_RTPSParticipant, this, RTPSMessageGroup::WRITER, m_cdrmessages, locatorsList, guids);
+        const LocatorList_t remote_locators_shrinked{mp_RTPSParticipant->network_factory().ShrinkLocatorLists(
+                {rp->m_att.endpoint.remoteLocatorList})};
```
Comment (MiguelCompany) [inline src/cpp/rtps/writer/StatefulWriter.cpp]: If I remeber correctly, remoteLocatorList is shrinked inside `matched_reader_add`, so there is no need to shrink it here

File: src/cpp/rtps/writer/StatefulWriter.cpp [hunk 19]
```diff
@@ -890,11 +889,12 @@ SequenceNumber_t StatefulWriter::next_sequence_number() const
 void StatefulWriter::send_heartbeat_to_nts(ReaderProxy& remoteReaderProxy, bool final)
 {
     std::vector<GUID_t> tmp_guids(1, remoteReaderProxy.m_att.guid);
-    const LocatorList_t& locators = remoteReaderProxy.m_att.endpoint.remoteLocatorList;
+    const LocatorList_t remote_locators_shrinked{mp_RTPSParticipant->network_factory().ShrinkLocatorLists(
+            {remoteReaderProxy.m_att.endpoint.remoteLocatorList})};
```
Comment (MiguelCompany) [inline src/cpp/rtps/writer/StatefulWriter.cpp]: If I remeber correctly, remoteLocatorList is shrinked inside `matched_reader_add`, so there is no need to shrink it here

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1229/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1199/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1212/)

File: include/fastrtps/rtps/attributes/ReaderAttributes.h [hunk 20]
```diff
@@ -37,8 +37,8 @@ class ReaderTimes
 public:
     ReaderTimes()
     {
-        initialAcknackDelay.fraction = 200*1000*1000;
-        heartbeatResponseDelay.fraction = 20*1000*1000;
+        initialAcknackDelay.fraction = 300*1000*1000;
+        heartbeatResponseDelay.fraction = 50*1000*1000;
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastrtps/rtps/attributes/ReaderAttributes.h]: Done. Changes only applied to SEDP endpoints.

File: include/fastrtps/rtps/attributes/WriterAttributes.h [hunk 21]
```diff
@@ -45,9 +45,10 @@ class  WriterTimes
 public:
     WriterTimes()
     {
-        initialHeartbeatDelay.fraction = 200*1000*1000;
+        initialHeartbeatDelay.fraction = 50*1000*1000;
         heartbeatPeriod.seconds = 3;
-        nackResponseDelay.fraction = 20*1000*1000;
+        nackResponseDelay.fraction = 400*1000*1000;
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastrtps/rtps/attributes/WriterAttributes.h]: Done. Changes only applied to SEDP endpoints.

File: include/fastrtps/rtps/writer/timedevent/NackResponseDelay.h (hunk 1 above)
Comment (richiware) [reply include/fastrtps/rtps/writer/timedevent/NackResponseDelay.h]: Done

File: include/fastrtps/rtps/writer/timedevent/NackResponseDelay.h (hunk 2 above)
Comment (richiware) [reply include/fastrtps/rtps/writer/timedevent/NackResponseDelay.h]: Done

File: src/cpp/rtps/writer/StatefulWriter.cpp (hunk 15 above)
Comment (richiware) [reply src/cpp/rtps/writer/StatefulWriter.cpp]: Done

File: src/cpp/rtps/writer/StatefulWriter.cpp (hunk 16 above)
Comment (richiware) [reply src/cpp/rtps/writer/StatefulWriter.cpp]: Decided by the team to keep in this way.

File: src/cpp/rtps/writer/StatefulWriter.cpp (hunk 17 above)
Comment (richiware) [reply src/cpp/rtps/writer/StatefulWriter.cpp]: Decided by the team to keep in this way.

File: src/cpp/rtps/writer/StatefulWriter.cpp (hunk 18 above)
Comment (richiware) [reply src/cpp/rtps/writer/StatefulWriter.cpp]: Decided by the team to keep in this way.

File: src/cpp/rtps/writer/StatefulWriter.cpp (hunk 19 above)
Comment (richiware) [reply src/cpp/rtps/writer/StatefulWriter.cpp]: Decided by the team to keep in this way.

File: src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp (hunk 4 above)
Comment (richiware) [reply src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp]: Done

File: src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp (hunk 5 above)
Comment (richiware) [reply src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp]: Done

File: src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp (hunk 6 above)
Comment (richiware) [reply src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp]: Done

File: src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp (hunk 7 above)
Comment (richiware) [reply src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp]: Done

File: src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp (hunk 8 above)
Comment (richiware) [reply src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp]: Done

File: src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp (hunk 9 above)
Comment (richiware) [reply src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp]: Done

File: src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp (hunk 10 above)
Comment (richiware) [reply src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp]: Done

File: src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp (hunk 11 above)
Comment (richiware) [reply src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp]: Done

File: src/cpp/rtps/reader/timedevent/InitialAckNack.cpp (hunk 12 above)
Comment (richiware) [reply src/cpp/rtps/reader/timedevent/InitialAckNack.cpp]: Done

File: src/cpp/rtps/writer/ReaderProxy.cpp (hunk 13 above)
Comment (richiware) [reply src/cpp/rtps/writer/ReaderProxy.cpp]: Done

File: src/cpp/rtps/writer/StatefulWriter.cpp (hunk 14 above)
Comment (richiware) [reply src/cpp/rtps/writer/StatefulWriter.cpp]: Done

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1236/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1206/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1219/)

**Comment by clalancette** [comment]
I did some additional testing with this PR yesterday, and it certainly seems to improve the situation.  In the ROS2 application I'm working on, there about 2 dozen topics publishing at around 20Hz.  Before this patch, it could take up to 30 seconds to do discovery.  Afterwards, the worst case I saw was about 2 seconds, though the mean time for discovery was in the 10s-of-milliseconds range. 

We may still have to do some debugging on those long outliers (particularly since everything is localhost), but this is a huge improvement.  Thanks!

**Comment by sloretz** [comment]
@clalancette, @richiware, @MiguelCompany It looks like a couple ROS 2 tests are failing after this PR was merged (See ros2/build_cop#166, Tests pass in bc61fa3fab63ced48416db86ebb8df4e0f0cebdf but fail in cb0a6e237eee4985ef8b1ffe6afbebd8e4116d04 ). Specifically the tests are timing out. What information should I gather to help debug?

Failing tests
https://github.com/ros2/rclcpp/blob/8743bcb0a1ef06d6eea5ebbd247e019f9dbb1e10/rclcpp_action/test/test_client.cpp#L269-L288
https://github.com/ros2/system_tests/blob/master/test_cli/test/test_params_yaml.py

**Pull request merged**